File: 0107.binary_tree_level_order_traversal_II/solution.java
import java.util.LinkedList;

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        LinkedList<List<Integer>> ans = new LinkedList<List<Integer>>();
        addLevel(root, 0, ans);
        return ans;
    }

    public void addLevel(TreeNode node, int level, LinkedList<List<Integer>> list){
        if (node == null){
            return;
        }
        if (level+1 > list.size()){
            list.addFirst(new LinkedList<Integer>());
        }
        list.get(list.size()-1-level).add(node.val);
        addLevel(node.left, level+1, list);
        addLevel(node.right, level+1, list);
    }
}

File: 0107.binary_tree_level_order_traversal_II/solution.py
from typing import List


from utils import TreeNode


class Solution:
    def level_order_bottom(self, root: TreeNode) -> List[List[int]]:
        """ use stack """
        stack = [(root, 0)]
        ans = []
        while stack:
            node, level = stack.pop()
            if node:
                if len(ans) < level + 1:
                    ans.append([])
                ans[level].append(node.val)
                stack.append((node.right, level + 1))
                stack.append((node.left, level + 1))
        return ans[::-1]
